b2sdk/bucket.py

```diff
@@ -354,6 +359,32 @@ def ls(
         if prefix != '' and not prefix.endswith('/'):
             prefix += '/'
 
+        # If we're running with wildcard-matching, we could get
+        # a different prefix from it.  We search for the first
+        # occurrence of any of the special characters and fetch
+        # parent path from that place.
+        # Examples:
+        #   'b/c/*.txt' –> 'b/c/'
+        #   '*.txt' –> ''
+        #   'a/*/result.[ct]sv' –> 'a/'
+        if with_wildcard:
+            for wildcard_character in ['*', '?', '[']:
+                if wildcard_character not in folder_to_list:
+                    continue
+
+                starter_index = folder_to_list.index(wildcard_character)
+                # +1 to include the starter character.  Using posix path to
+                # ensure consistent behaviour on Windows (e.g. case sensitivity).
+                path = pathlib.PurePosixPath(folder_to_list[:starter_index + 1])
+                parent_path = str(path.parent)
+                # Path considers dot to be the empty path.
+                # There's no shorter path than that.
+                if parent_path == '.':
+                    prefix = ''
+                    break
+                if len(parent_path) < len(prefix):
```

Comment:
when, for example, that happens?

Reply:
With `folder_to_list` being e.g. `'a/[ab]/*.csv'` you'll have the initial prefix of `'a/[ab]/*.csv/'` (nonsensical), then `'a/[ab]/'` (better) and finally `'a/'` – the correct one.
In case of `'a/*/test.[ct]vs'` you go to `'a/'` first and to `'a/*/'` on third iteration. And the `'a/'` is the correct one that you want to start from.

Reply:
I've added an additional comment for the future readers.